File: src/java/DigestTools.java
package java;

import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

class DigestTools {
    static byte[] getDigest(BigInteger bigIntValue, String...
            instanceAlgo) {
        if (instanceAlgo[0].isEmpty()) instanceAlgo[0] = "SHA-256";
        String biString = bigIntValue.toString();
        MessageDigest digest = null;
        try {
            digest = MessageDigest.getInstance(instanceAlgo[0]);
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        }
        return digest.digest(biString.getBytes(StandardCharsets
                .UTF_8));
    }

    static void digestPrinter(String digestHex) {
        // left padding for odd-length strings
        if (digestHex.length() % 2 != 0) digestHex = " " + digestHex;
        for (int i = 0; i < digestHex.length(); i += 2) {
            char currentChar = digestHex.charAt(i);
            char nextChar = digestHex.charAt(i+1);
            System.out.printf("%c%c ", currentChar, nextChar);
        }
    }
}

File: src/kotlin/DigestTools.kt
package kotlin

import java.math.BigInteger
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException

internal object DigestTools {
    fun getDigest(bigIntValue: BigInteger, instance: String?):
            ByteArray {
        var instanceAlgo: String? = instance
        if (instanceAlgo.isNullOrEmpty()) instanceAlgo = "SHA-256"
        val biString = bigIntValue.toString()
        var digest: MessageDigest? = null
        try {
            digest = MessageDigest.getInstance(instanceAlgo)
        } catch (e: NoSuchAlgorithmException) {
            e.printStackTrace()
        }

        return digest!!.digest(biString.toByteArray(StandardCharsets
                .UTF_8))
    }

    fun digestPrinter(digestHex: String) {
        var digestHex = digestHex
        // left padding for odd-length strings
        if (digestHex.length % 2 != 0) digestHex = " " + digestHex
        var i = 0
        while (i < digestHex.length) {
            val currentChar = digestHex[i]
            val nextChar = digestHex[i + 1]
            System.out.printf("%c%c ", currentChar, nextChar)
            i += 2
        }
    }
}
